### PR TITLE
Improve template integration tests

### LINKED ISSUE #562: Avoid quadratic scaling of template integration tests
#### Issue description

Currently, in ``test_templates.py`` a compatibility test is performed by having every template in the library (of the same device) applied before every other one - a quadratic growth of test cases in the number of templates. This becomes prohibitive, and we should find another solution that tests templates' compatibility.

#### Additional information

The issue could be easily fixed by defining small dummy templates that are called before and after the tested template. We could also try to make the tested templates pass parameters to each other.

### PR DESCRIPTION
**Pull request by mariaschuld**
**Context:** 

To facilitate the addition of new templates, the integration tests needed a major overhaul to be simplified.

**Description of the Change:**

* Now there are only two fixtures (or better: list of fixture entries ``(..., ..., )``) to which templates have to be added. One defines which tests to run for the initialization function integration tests, the other defines which tests to run for all others.
* Templates that need to be excluded from certain tests are handed manually
* Tests of qnode integration and integration with other quantum ops are separated
* Testing templates with other quantum ops now uses a simple gate, instead of another template

**Related GitHub Issues:** Fixes #562

### CODE REVIEW

**Comment by mariaschuld** [comment]
Apologies for the ugly diff file. It was a major overhaul...

**Comment by josh146** [comment]
Does this also address any of the issues in #421?

**Comment by mariaschuld** [comment]
Unfortunately not @josh146 . I did check, and the issue is mostly about the other tests...

File: tests/templates/test_integration.py [hunk 1]
```diff
@@ -176,558 +138,331 @@
...
-class TestIntegrationCircuit:
-    """Tests the integration of templates into circuits using different interfaces. """
-
-    @pytest.mark.parametrize("template1, diffable1, nondiffable1", QUBIT_DIFFABLE_NONDIFFABLE)
-    @pytest.mark.parametrize("template2, diffable2, nondiffable2", QUBIT_DIFFABLE_NONDIFFABLE)
+# List templates in NO_OP_BEFORE that do not allow for operations
+# before they are called in a quantum function.
+# These templates will be skipped in tests of that nature.
+
+NO_OP_BEFORE = [qml.templates.AmplitudeEmbedding, qml.templates.BasisEmbedding]
```
Comment (antalszava) [inline tests/templates/test_integration.py:152]: Out of curiosity: shouldn't state preparations (e.g. BasisState and Mottonen) be included here? Also curious about the other types of embeddings.

File: tests/templates/test_integration.py [hunk 2]
```diff
@@ -176,558 +138,331 @@
...
+              (qml.templates.StronglyEntanglingLayers,
+               {'weights': qml.init.strong_ent_layers_normal(n_layers=2, n_wires=3), 'wires': range(3)}),
+              (qml.templates.RandomLayers,
+               {'weights': qml.init.random_layers_uniform(n_layers=3, n_rots=2, n_wires=2), 'wires': range(2)}),
+              (qml.templates.RandomLayers,
+               {'weights': qml.init.random_layers_uniform(n_layers=3, n_rots=2, n_wires=2), 'wires': range(2)}),
+              (qml.templates.RandomLayers,
+               {'weights': qml.init.random_layers_normal(n_layers=2, n_rots=2, n_wires=3), 'wires': range(3)}),
+              (qml.templates.RandomLayers,
+               {'weights': qml.init.random_layers_normal(n_layers=2, n_rots=2, n_wires=3), 'wires': range(3)}),
```
Comment (antalszava) [inline tests/templates/test_integration.py:178]: There seems to be a duplication here (2-2 being pairwise duplicates), is that intentional?

File: tests/templates/test_integration.py [hunk 3]
```diff
@@ -176,558 +138,331 @@
...
+              (qml.templates.BasicEntanglerLayers,
+               {'weights': qml.init.basic_entangler_layers_normal(n_layers=3, n_wires=1), 'wires': range(1)}),
+              (qml.templates.BasicEntanglerLayers,
+               {'weights': qml.init.basic_entangler_layers_normal(n_layers=3, n_wires=2), 'wires': range(2)}),
+              (qml.templates.BasicEntanglerLayers,
+               {'weights': qml.init.basic_entangler_layers_normal(n_layers=3, n_wires=3), 'wires': range(3)}),
+              ]
+
+CV_INIT = [(qml.templates.CVNeuralNetLayers,
+            {'theta_1': qml.init.cvqnn_layers_theta_uniform(n_layers=3, n_wires=2),
```
Comment (antalszava) [inline tests/templates/test_integration.py:222]: FYI: each entry appears twice in here as well

File: tests/templates/test_integration.py [hunk 4]
```diff
@@ -12,88 +12,29 @@
...
+* Try to import the interface and add its variable creation function to INTERFACES
 * Extend the fixture ``interfaces``
 * Add the interface gradient computation to the TestGradientIntegration tests
-
 """
 # pylint: disable=protected-access,cell-var-from-loop
 import pytest
 import numpy as np
 import pennylane as qml
-from pennylane.templates import (Interferometer,
```
Comment (antalszava) [inline tests/templates/test_integration.py:45]: Nice that these can be removed! :)

File: tests/templates/test_integration.py [hunk 5]
```diff
@@ -176,558 +138,331 @@
...
 
         # Extract keys and items
-        keys_diffable1 = [*diffable1]
-        diffable1 = list(diffable1.values())
-        keys_diffable2 = [*diffable2]
-        diffable2 = list(diffable2.values())
-
-        # Combine diffable inputs to allow passing with *
-        diffable = diffable1 + diffable2
+        keys_diffable = [*diffable]
```
Comment (antalszava) [inline tests/templates/test_integration.py:281]: Curious about the diff here: this seems to be addressing #562, if I understand correctly now only one template at a time is being tested, correct? Was there any reason (e.g. edge case) for testing each template pairwise before?

File: tests/templates/test_integration.py (hunk 1 above)
Comment (mariaschuld) [reply tests/templates/test_integration.py:152]: Very good question!

Comment (mariaschuld) [reply tests/templates/test_integration.py:152]: That was quite involved:

- ``MottonenStatePrep`` can be used with preceding operations
- ``BasisState`` is not a template but an ``op``, so not explicitly tested here
- ``BasisStatePreparation`` *is* a template and tested, but implemented by PauliX rotations and can therefore be used with preceding operations
- ``AmplitudeEmbedding`` uses the op ``QubitStateVec`` on ``default.qubit`` and cannot be used with preceding operations
- ``BasisEmbedding`` was using the op ``BasisState``, but I changed it to be implemented with PauliX rotations, so this edge case disappears

Your comment uncovered a bug in the tests - using the type checking works for ``ops`` but not for ``templates``, since they are functions. So checking for ``NO_OP_BEFORE`` accidentally excluded *all* templates from that test case. This is fixed now by instead checking the name of the function.

File: tests/templates/test_integration.py (hunk 2 above)
Comment (mariaschuld) [reply tests/templates/test_integration.py:178]: Well spotted!!

I think that was a classic "copy and paste, but forgot to change". Fixed.

File: tests/templates/test_integration.py (hunk 3 above)
Comment (mariaschuld) [reply tests/templates/test_integration.py:222]: Thanks! I wanted the duplicates to test the 1-wire edge case!

File: tests/templates/test_integration.py (hunk 5 above)
Comment (mariaschuld) [reply tests/templates/test_integration.py:281]: Not at all, I just thought it was maximally diversifying. 

What I could do is use a hand-made template in the integration test, in addition to the gate.

Comment (mariaschuld) [reply tests/templates/test_integration.py:281]: Implemented that!

**Comment by mariaschuld** [comment]
@antalszava, your comments were great! I should have addressed them all.

File: pennylane/templates/embeddings/basis.py [hunk 6]
```diff
@@ -15,12 +15,11 @@
 Contains the ``BasisEmbedding`` template.
```
Comment (mariaschuld) [inline pennylane/templates/embeddings/basis.py:15]: Sorry, I snuck in this overdue change, which makes ``BasisEmbedding`` independent from ``BasisState``.

File: tests/templates/test_embeddings.py [hunk 7]
```diff
@@ -329,6 +329,21 @@ def circuit(x=None):
         with pytest.raises(ValueError, match="'basis_state' must only consist of"):
             circuit(x=[2, 3])
```
Comment (mariaschuld) [inline tests/templates/test_embeddings.py:331]: Also snuck in this test of the additional check in ``BasisEmbedding``

File: tests/templates/test_integration.py (hunk 1 above)
Comment (antalszava) [reply tests/templates/test_integration.py]: Oh, that's really nice, thanks for the details! :blush: :1st_place_medal:

Comment (antalszava) [reply tests/templates/test_integration.py]: (A follow-up for a final clarification): these descriptions are about how each can be used inside of a `qfunc` functionally. Would it still make sense to (at one point, definitely out of scope here) add a custom error to each of them if they are not the first operation in the `qfunc`?

File: tests/templates/test_integration.py [hunk 8]
```diff
@@ -176,558 +138,347 @@
...
+               {'features': [1., 2.], 'weights': qml.init.qaoa_embedding_uniform(n_layers=3, n_wires=2),
+                'wires': range(2)}),
+              (qml.templates.QAOAEmbedding,
+               {'features': [1.], 'weights': qml.init.qaoa_embedding_normal(n_layers=2, n_wires=1),
+                'wires': range(1)}),
+              (qml.templates.QAOAEmbedding,
+               {'features': [1., 2.], 'weights': qml.init.qaoa_embedding_normal(n_layers=2, n_wires=2),
+                'wires': range(2)}),
+              (qml.templates.QAOAEmbedding,
+               {'features': [1.], 'weights': qml.init.qaoa_embedding_normal(n_layers=2, n_wires=1),
```
Comment (antalszava) [inline tests/templates/test_integration.py:192]: Seems to be the same as on lines 185-187
```suggestion
```

File: tests/templates/test_integration.py [hunk 9]
```diff
@@ -176,558 +138,347 @@
...
 
         # Turn into correct format
         diffable = [to_var(i) for i in diffable]
 
         # Generate qnode
-        circuit = qnode_cv_args(gaussian_device_2_wires, interface, template1, template2, len(diffable1))
+        @qml.qnode(gaussian_device_2_wires, interface=interface)
+        def circuit(*diffable, keys_diffable=None, nondiffable=None):
+            # Turn diffables back into dictionary
+            dict = {key: item for key, item in zip(keys_diffable, diffable)}
```
Comment (antalszava) [inline tests/templates/test_integration.py:324]: (Minor suggestion): this also seems to work
```suggestion
            dict = dict(zip(keys_diffable, diffable))
```

File: tests/templates/test_integration.py (hunk 1 above)
Comment (mariaschuld) [reply tests/templates/test_integration.py]: Definitely. And here it will be useful to have a template class, see also #574, #575...

File: tests/templates/test_integration.py [hunk 10]
```diff
@@ -176,558 +138,347 @@
...
-                                           interface, to_var):
-        """Tests integration of qubit templates passing differentiable arguments as keyword arguments to qnode."""
+    def test_qubit_qnode_auxiliary_args(self, template, diffable, nondiffable, interface, to_var):
+        """Tests integration of qubit templates passing differentiable arguments as auxiliary arguments to qnode."""
 
         # Change type of differentiable arguments
-        # TODO: templates should all take arrays AND lists, at the moment this is not the case
-        diffable1 = {k: np.array(v) for k, v in diffable1.items()}
-        diffable2 = {k: np.array(v) for k, v in diffable2.items()}
+        # Fix: templates should all take arrays AND lists, at the moment this is not the case
```
Comment (antalszava) [inline tests/templates/test_integration.py:344]: Is it the case that some take only arrays and some take only lists?

Comment (mariaschuld) [reply tests/templates/test_integration.py:344]: Yes, I think BasisStatePrep required an array...we should investigate this systematically, allow for both and add a test to each template which uses lists or arrays for the diffable arguments!

File: tests/templates/test_integration.py [hunk 11]
```diff
@@ -176,558 +138,347 @@
...
 
-    @pytest.mark.parametrize("template1, diffable1, nondiffable1", QUBIT_DIFFABLE_NONDIFFABLE)
-    @pytest.mark.parametrize("template2, diffable2, nondiffable2", QUBIT_DIFFABLE_NONDIFFABLE)
+    @pytest.mark.parametrize("template, diffable, nondiffable", QUBIT_DIFFABLE_NONDIFFABLE)
     @pytest.mark.parametrize("interface, to_var", INTERFACES)
-    def test_integration_qubit_nondiffable(self, template1, diffable1, nondiffable1,
-                                           template2, diffable2, nondiffable2,
-                                           interface, to_var):
-        """Tests integration of qubit templates passing differentiable arguments as keyword arguments to qnode."""
+    def test_qubit_qnode_auxiliary_args(self, template, diffable, nondiffable, interface, to_var):
```
[the remark this answers is not in the document]
Comment (trbromley) [reply tests/templates/test_integration.py:340]: I think I didn't follow the motivation for the auxiliary args tests

File: tests/templates/test_integration.py [hunk 12]
```diff
@@ -176,558 +138,347 @@
...
+        circuit(nondiffable=nondiffable)
 
-    @pytest.mark.parametrize("template, dict", CV_INIT)
-    def test_integration_qubit_init(self, template, dict, gaussian_dummy):
-        """Checks parameter initialization compatible with qubit templates."""
+    @pytest.mark.parametrize("op_before_template", [True, False])
+    @pytest.mark.parametrize("template, diffable, nondiffable", CV_DIFFABLE_NONDIFFABLE)
+    def test_integration_cv_nondiffable(self, template, diffable, nondiffable, gaussian_device_2_wires,
+                                        op_before_template):
+        """Tests integration of cv templates passing differentiable arguments as auxiliary arguments to qnode."""
```
Comment (trbromley) [inline tests/templates/test_integration.py:446]: Need to update name to be like the qubit one? "test_cv_template_followed_by_operation"
And also the test docstring.

File: tests/templates/test_integration.py [hunk 13]
```diff
@@ -176,558 +138,347 @@
...
 
-        n_wires = len(dict['wires'])
-        dev = gaussian_dummy(n_wires)
+        # Change type of differentiable arguments
```
Comment (trbromley) [inline tests/templates/test_integration.py:448]:
```suggestion
        # skip this test if template does not allow for operations before
        if template.__name__ in NO_OP_BEFORE and op_before_template:
            pytest.skip("Template does not allow operations before - skipping this test.")
            
        # Change type of differentiable arguments
```
Copying from the qubit case? I guess now there aren't any CV templates in `NO_OP_BEFORE`, but maybe there will be in future?

File: tests/templates/test_integration.py [hunk 14]
```diff
@@ -176,558 +138,347 @@
...
         diffable = [to_var(i) for i in diffable]
 
         # Generate qnode
         dev = qml.device('default.qubit', wires=2)
-        circuit = qnode_qubit_args(dev, interface, template1, template2, len(diffable1))
+
+        @qml.qnode(dev, interface=interface)
+        def circuit(*diffable, keys_diffable=None, nondiffable=None):
+            # Turn diffables back into dictionary
+            dict = {key: item for key, item in zip(keys_diffable, diffable)}
```
Comment (trbromley) [inline tests/templates/test_integration.py:293]: Is overwriting `dict` problematic?

File: tests/templates/test_integration.py [hunk 15]
```diff
@@ -754,30 +505,30 @@ def circuit(*diffable):
...
-                tape.gradient(loss, grad_inpts)
-
-    @pytest.mark.parametrize("template, diffable, nondiffable, argnum", CV_DIFFABLE_NONDIFFABLE_ARGNUM)
+        # Do gradient check for every differentiable argument
+        for argnum in range(len(diffable)):
+
+            # Check gradients in numpy interface
+            if interface == 'numpy':
+                grd = qml.grad(circuit, argnum=[argnum])
+                grd(*diffable)
```
Comment (trbromley) [inline tests/templates/test_integration.py:514]: I printed the gradient and got zeros everywhere, which is expected if we're taking the expectation value w.r.t. the identity. But I also changed the observable to `PauliZ` and still got all zeros. Is that right? :thinking:

I'm just wondering if there's something stronger that we can say, e.g., that at least one gradient is nonzero?

File: tests/templates/test_integration.py (hunk 14 above)
Comment (trbromley) [reply tests/templates/test_integration.py:293]: I think this is why the tests are failing.

File: tests/templates/test_embeddings.py (hunk 7 above)
Comment (mariaschuld) [reply tests/templates/test_embeddings.py:331]: Yes, it is the same thing, just in a different packaging and with different input checks :) But not all users may know this, so the templates may be used in other contexts.

I could make one dependent on the other, but find it more elegant to implement the 2 lines of code from scratch, for better maintainability.

``BasisState`` "hacks" the state preparation on simulators as far as I know.

File: tests/templates/test_integration.py [hunk 16]
```diff
@@ -176,558 +138,347 @@
...
+                'wires': range(2)}),
+              (qml.templates.QAOAEmbedding,
+               {'features': [1.], 'weights': qml.init.qaoa_embedding_normal(n_layers=2, n_wires=1),
+                'wires': range(1)}),
+              (qml.templates.QAOAEmbedding,
+               {'features': [1., 2.], 'weights': qml.init.qaoa_embedding_normal(n_layers=2, n_wires=2),
+                'wires': range(2)}),
+              (qml.templates.QAOAEmbedding,
+               {'features': [1.], 'weights': qml.init.qaoa_embedding_normal(n_layers=2, n_wires=1),
+                'wires': range(1)}),
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply tests/templates/test_integration.py:193]: Ah man, I thought I caught them all...thanks!

File: tests/templates/test_integration.py [hunk 17]
```diff
@@ -176,558 +138,347 @@
...
+              (qml.templates.QAOAEmbedding,
+               {'features': [1., 2.], 'weights': qml.init.qaoa_embedding_uniform(n_layers=2, n_wires=2),
+                'wires': range(2)}),
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply tests/templates/test_integration.py:196]: Good point!

File: tests/templates/test_integration.py [hunk 18]
```diff
@@ -176,558 +138,347 @@
...
+# These templates will be skipped in tests of that nature.
+
+NO_OP_BEFORE = ["AmplitudeEmbedding"]
+
+# Each entry to QUBIT_INIT and CV_INIT adds a template with specified inputs to the
+# integration tests ``TestIntegrationInitFunctions``
+
+# The entries have the following form:
+#
+# (template, dict of differentiable arguments, dict of non-differentiable arguments)
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply tests/templates/test_integration.py:159]: Another good find!

File: tests/templates/test_integration.py [hunk 19]
```diff
@@ -176,558 +138,347 @@
...
+                'wires': range(2)}),
+              (qml.templates.SimplifiedTwoDesign,
+               {'initial_layer_weights': qml.init.simplified_two_design_initial_layer_uniform(n_wires=4),
+                'weights': qml.init.simplified_two_design_weights_uniform(n_layers=3, n_wires=4),
+                'wires': range(4)}),
+              (qml.templates.SimplifiedTwoDesign,
+               {'initial_layer_weights': qml.init.simplified_two_design_initial_layer_normal(n_wires=4),
+                'weights': qml.init.simplified_two_design_weights_normal(n_layers=3, n_wires=4),
+                'wires': range(4)}),
+              (qml.templates.BasicEntanglerLayers,
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply tests/templates/test_integration.py:205]: The FIXTURE dictionary could include any (non-diff or diff) arguments. But since I want to test the init functions, I take all keyword arguments with their default values. The ``rot`` argument should be tested in the template's unit test class elsewhere...

File: tests/templates/test_integration.py [hunk 20]
```diff
@@ -124,46 +65,67 @@
...
+                              (qml.templates.MottonenStatePreparation,
+                               {'state_vector': np.array([1 / 2, 1 / 2, 1 / 2, 1 / 2])},
+                               {}),
+                              (qml.templates.BasisStatePreparation,
+                               {},
+                               {'basis_state': np.array([1, 0])}),
+                              (qml.templates.StronglyEntanglingLayers,
                                {'weights': [[[4.54, 4.79, 2.98], [4.93, 4.11, 5.58]],
                                             [[6.08, 5.94, 0.05], [2.44, 5.07, 0.95]]]},
                                {}),
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply tests/templates/test_integration.py:98]: It's not 100 strict - you can test here whatever you want in terms of integration. My thinking is that we reduce the test cases here by using defaults wherever possible - the keyword args should be tested in the "non"-integration tests (i.e. ``test_layers``)...

File: tests/templates/test_integration.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mariaschuld) [reply tests/templates/test_integration.py:340]: Yes, we seem to have introduced the convention of naming positional args of qnodes "primary" and keyword args "auxiliary". Primary args are those with respect to which we can compute gradients. 

The tests have the reason that the way we pass things to the qnode influences what shape the argument has inside the qnode, when the template calls it.

Added explanation...

File: tests/templates/test_integration.py [hunk 21]
```diff
@@ -176,558 +138,347 @@
...
-        """Tests integration of qubit templates passing differentiable arguments as keyword arguments to qnode."""
+    def test_qubit_qnode_auxiliary_args(self, template, diffable, nondiffable, interface, to_var):
+        """Tests integration of qubit templates passing differentiable arguments as auxiliary arguments to qnode."""
 
         # Change type of differentiable arguments
-        # TODO: templates should all take arrays AND lists, at the moment this is not the case
-        diffable1 = {k: np.array(v) for k, v in diffable1.items()}
-        diffable2 = {k: np.array(v) for k, v in diffable2.items()}
+        # Fix: templates should all take arrays AND lists, at the moment this is not the case
+        diffable = {k: np.array(v) for k, v in diffable.items()}
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply tests/templates/test_integration.py:345]: Yes, and I hope this can be fixed soon: Some templates don't like lists here...

File: tests/templates/test_integration.py (hunk 13 above)
Comment (mariaschuld) [reply tests/templates/test_integration.py:448]: I'd prefer if people who add tests adapt the structure to their needs...

File: tests/templates/test_integration.py (hunk 14 above)
Comment (mariaschuld) [reply tests/templates/test_integration.py]: That is a very good point - it seemed to work but with the last changes it might have stopped. Anyways, better to be safe!

File: tests/templates/test_integration.py (hunk 15 above)
Comment (mariaschuld) [reply tests/templates/test_integration.py:534]: Depends on which case you check - for me AngleEmbedding and numpy were giving nice non-zero gradients...

**Comment by mariaschuld** [comment]
@trbromley  thanks for your extremely useful comments. If you like to re-review, I should have implemented all your changes.

File: tests/templates/test_integration.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply tests/templates/test_integration.py:101]: Excellent, makes sense

File: tests/templates/test_integration.py [hunk 22]
```diff
@@ -75,6 +75,9 @@
 # The entries have the following form:
 # (template, dict of differentiable arguments, dict of non-differentiable arguments)
 
+# "Differentiable arguments" to a template are those that in principle allow a user to compute gradients for,
```
Comment (trbromley) [inline tests/templates/test_integration.py:78]:
```suggestion
# "Differentiable arguments" in a template are those that in principle allow a user to compute gradients for,
```

File: tests/templates/test_integration.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply tests/templates/test_integration.py:356]: Thanks!

File: tests/templates/test_integration.py [hunk 23]
```diff
@@ -269,7 +273,16 @@
 
 class TestIntegrationQnode:
     """Tests the integration of templates into qnodes when differentiable arguments are passed as
-    primary or auxiliary arguments to the qnode, using different interfaces."""
+    primary or auxiliary arguments to the qnode, using different interfaces.
+
+    "Differentiable arguments" to a template are those that in principle allow a user to compute gradients for,
```
Comment (trbromley) [inline tests/templates/test_integration.py:278]:
```suggestion
    "Differentiable arguments" in a template are those that in principle allow a user to compute gradients for,
```
Actually I'm not sure, maybe the original version is fine too :thinking:

File: tests/templates/test_integration.py (hunk 13 above)
Comment (trbromley) [reply tests/templates/test_integration.py:468]: Ok, up to you, I think it's good to minimize the overhead - I guess it depends how likely we expect a CV template to come in soon that doesn't allow operations before it.
